multipass: 1.14.0 -> 1.15.0

jnsgruk opened the pull request:
This commit is a fairly big one:

- Package up the new [1.15.0-rc4](https://github.com/canonical/multipass/releases/tag/v1.15.0-rc4) release
- Include the Flutter desktop application
- Move to `pkgs/by-name`

I've pushed this a little early while we're still in the late `rc` phase of the Multipass release to get feedback on the restructuring. I noted the following in the commit that introduces the GUI element:

    This commit breaks down multipass into two intermediate packages,
    `multipassd` and `multipass-gui`. These are then packaged up with
    `symlinkJoin` to work around some intricacies in the
    `buildFlutterApplication` process which would otherwise be difficult
    to patch.
    
    The multipass repo is setup upstream such that the flutter app would
    be built as part of the regular cmake process, but that can't work
    here due to the fetching of the flutter dependencies. The upstream
    multipass project builds `libdart_ffi.so`, which is needed in the
    LD_LIBRARY_PATH of the GUI. Building the two derivations seperately
    enables that to be done relatively simply.

Hopefully that makes sense! Tested locally on my machine and all works as expected.

Perhaps @RossComputerGuy could give some feedback on this approach, seems they have lots of involvement with the Flutter tooling in nixpkgs.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jnsgruk commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 363626`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>multipass</li>
  </ul>
</details>

getchoo commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,75 @@
+
+  extraWrapProgramArgs = ''
+    --prefix LD_LIBRARY_PATH : '${multipassd}/lib'
+  '';
```
```suggestion
  '';
  
  meta = {
    inherit (meta) homepage license maintainers platforms;
  };
```
While much more annoying, this keeps attributes like `meta.pos` intact for use with `nix edit`, etc

jnsgruk replied:
Fixed, thanks :)

getchoo commented on pkgs/by-name/mu/multipass/multipassd.nix:
```diff
@@ -1,62 +1,48 @@
-      rm -rf .git
-    '';
-  };
+  inherit version meta;
```
See above

RossComputerGuy replied:
Should be fine like this, afaik meta.position only changes when there's an explicit attribute, so only if that meta attrset has it would it change.

jnsgruk replied:
Fixed, thanks :)

getchoo commented on pkgs/by-name/mu/multipass/package.nix:
```diff
@@ -0,0 +1,64 @@
+    multipass = nixosTests.multipass;
+  };
+
+  meta = commonMeta // {
```
Ditto

jnsgruk replied:
Fixed, thanks :)

getchoo commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,75 @@
+    mkdir -p lib/generated
+
+    # Generate the Dart gRPC code for the Multipass GUI.
+    ${protobuf}/bin/protoc \
```
```suggestion
    ${lib.getExe protobuf} \
```

jnsgruk replied:
Fixed, thanks :)

getchoo commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,75 @@
+    mv $out/bin/multipass_gui $out/bin/multipass.gui
+
+    substituteInPlace $out/share/applications/multipass.gui.autostart.desktop \
+      --replace-fail "Exec=multipass.gui" "Exec=$out/bin/multipass.gui" \
```
```suggestion
```
https://github.com/NixOS/nixpkgs/issues/308324

jnsgruk replied:
Fixed, thanks :)

getchoo commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,75 @@
+
+    substituteInPlace $out/share/applications/multipass.gui.autostart.desktop \
+      --replace-fail "Exec=multipass.gui" "Exec=$out/bin/multipass.gui" \
+      --replace-fail "Icon=multipass.gui" "Exec=$out/app/multipass-gui/data/flutter_assets/assets/icon.png" \
```
```suggestion
```

Fairly sure this is a similar case

jnsgruk replied:
Fixed, thanks :)

getchoo commented on pkgs/by-name/mu/multipass/package.nix:
```diff
@@ -0,0 +1,64 @@
+    '';
+  };
+
+  commonMeta = with lib; {
+    homepage = "https://multipass.run";
+    license = licenses.gpl3Plus;
+    maintainers = with maintainers; [ jnsgruk ];
+    platforms = [ "x86_64-linux" ];
+  };
```
```suggestion
  commonMeta = {
    homepage = "https://multipass.run";
    license = lib.licenses.gpl3Plus;
    maintainers = with lib.maintainers; [ jnsgruk ];
    platforms = [ "x86_64-linux" ];
  };
```
Nit

jnsgruk replied:
Fixed, thanks :)

getchoo commented on pkgs/by-name/mu/multipass/package.nix:
```diff
@@ -0,0 +1,64 @@
+    multipass-gui
+  ];
+
+  passthru.tests = {
+    multipass = nixosTests.multipass;
+  };
```
```suggestion
  passthru.tests = lib.optionalAttrs stdenv.hostPlatform.isLinux {
    inherit (nixosTests) multipass;
  };
```
Making this conditional stops OfBorg from trying to evaluate this on incompatible platforms and always failing

jnsgruk replied:
Fixed, thanks :)

getchoo reviewed (changes requested):
Also is there a reason to be updating to a release candidate instead of waiting for a stable release?

jnsgruk commented:
> Also is there a reason to be updating to a release candidate instead of waiting for a stable release?

Thanks for the feedback, I'll make those changes today - and nice tip on the desktop file, i'd not spotted that before.

Indeed my intention was to wait until the 1.15.0 release, I meant to mark this as draft while i got feedback on the approach 🤦‍♂️

SuperSandro2000 commented on pkgs/by-name/mu/multipass/update.sh:
```diff
@@ -0,0 +1,86 @@
+
+set -euo pipefail
+
+multipass_dir="./pkgs/by-name/mu/multipass"
```
We should abduct this from $0

SuperSandro2000 commented on pkgs/by-name/mu/multipass/update.sh:
```diff
@@ -0,0 +1,86 @@
+}
+
+
+LATEST_TAG="$(curl -s https://api.github.com/repos/canonical/multipass/releases/latest | jq -r '.tag_name')"
```
This should optionally take a GITHUB_TOKEN. Please see other update scripts on how to do that.

SuperSandro2000 commented on pkgs/by-name/mu/multipass/package.nix:
```diff
@@ -0,0 +1,74 @@
+    updateScript = ./update.sh;
+  };
+
+  meta = {
+    description = "Ubuntu VMs on demand for any workstation";
+    inherit (commonMeta)
+      homepage
+      license
+      maintainers
+      platforms
+      ;
+  };
```
```suggestion
  meta = commonMeta // {
    description = "Ubuntu VMs on demand for any workstation";
  };
```
otherwise might forget inheriting new things

jnsgruk replied:
Hah, I just changed that back following the last review - which way around should I prefer :)

SuperSandro2000 commented on pkgs/by-name/mu/multipass/package.nix:
```diff
@@ -0,0 +1,74 @@
+    rev = "refs/tags/v${version}";
+    hash = "sha256-xwEM7Zsjkyi91oNyiX0i1mIJfBOoJ/4X47JBEqP8xYY=";
+    fetchSubmodules = true;
+    leaveDotGit = true;
+    postFetch = ''
+      # Workaround for https://github.com/NixOS/nixpkgs/issues/8567
+      cd $out
+      rm -rf .git
+    '';
```
```suggestion
    leaveDotGit = true;
    postFetch = ''
      # Workaround for https://github.com/NixOS/nixpkgs/issues/8567
      rm -rf $out/.git
    '';
```
but we do we even need leaveDotGit?

SuperSandro2000 commented on pkgs/by-name/mu/multipass/multipassd.nix:
```diff
@@ -157,15 +146,13 @@ stdenv.mkDerivation {
-    license = licenses.gpl3Plus;
-    maintainers = with maintainers; [ jnsgruk ];
-    platforms = [ "x86_64-linux" ];
+  meta = {
+    inherit (meta)
+      description
+      homepage
+      license
+      maintainers
+      platforms
+      ;
   };
```
same as below

SuperSandro2000 commented on pkgs/by-name/mu/multipass/multipassd.nix:
```diff
@@ -118,27 +107,27 @@ stdenv.mkDerivation {
 
   buildInputs = [
     gtest
+    kdePackages.qtbase
+    kdePackages.qtwayland
```
I don't think that is intended to be used like that. Please stay to the things under qt

jnsgruk replied:
The other option I see (for `qtbase`) is: `libsForQt5.qt5.qtbase` - which is a major release out of date, is that what this should switch to? :)

jnsgruk replied:
Okay, figured it out. Using `qt6.qtbase` etc now.

SuperSandro2000 commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,85 @@
+    --prefix LD_LIBRARY_PATH : '${multipassd}/lib'
+  '';
+
+  meta = {
+    inherit (meta)
+      description
+      homepage
+      license
+      maintainers
+      platforms
+      ;
+  };
```
Same as below

SuperSandro2000 commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,85 @@
+      $out/share/applications/multipass.gui.desktop
+  '';
+
+  extraWrapProgramArgs = ''
+    --prefix LD_LIBRARY_PATH : '${multipassd}/lib'
+  '';
```
Can we use patchelf instead? LD_LIBRARY_PATH is only a last resort if nothing can be used

SuperSandro2000 commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,85 @@
+    mkdir -p lib/generated
+
+    # Generate the Dart gRPC code for the Multipass GUI.
+    ${lib.getExe protobuf} \
```
Can't we add it to nativeBuildInputs?

SuperSandro2000 commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,85 @@
+    ${lib.getExe protobuf} \
+      --plugin=protoc-gen-dart=${lib.getExe protoc-gen-dart} \
+      --dart_out=grpc:lib/generated \
+      -I ${multipass_src}/src/rpc \
+      ${multipass_src}/src/rpc/multipass.proto \
```
Can't we use relative dirs?

jnsgruk committed:
multipass: add update script

jnsgruk committed:
multipass: 1.14.0 -> 1.15.0

jnsgruk committed:
multipass: build and install the flutter gui

This commit breaks down multipass into two intermediate packages,
`multipassd` and `multipass-gui`. These are then packaged up with
`symlinkJoin` to work around some intricacies in the
`buildFlutterApplication` process which would otherwise be difficult
to patch.

The multipass repo is setup upstream such that the flutter app would
be built as part of the regular cmake process, but that can't work
here due to the fetching of the flutter dependencies. The upstream
multipass project builds `libdart_ffi.so`, which is needed in the
LD_LIBRARY_PATH of the GUI. Building the two derivations seperately
enables that to be done relatively simply.

jnsgruk commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 363626`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>multipass</li>
  </ul>
</details>

jnsgruk commented:
@SuperSandro2000 @getchoo feedback addressed, and also now bumped to the actual release, not the RC. Hopefully good to go now! Thanks for the help.

jnsgruk commented:
And some screenshots 😄 

![image](https://github.com/user-attachments/assets/4b160a0b-9b62-48b1-beb2-527512bafe65)

![image](https://github.com/user-attachments/assets/624334e8-cd95-4fa0-b7c5-7a709239aa98)

![image](https://github.com/user-attachments/assets/64bc7a53-9c49-4e2c-aacc-2a0f7a3fb539)

SuperSandro2000 commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,79 @@
+    mkdir -p lib/generated
+
+    # Generate the Dart gRPC code for the Multipass GUI.
+    ${lib.getExe protobuf} \
```
```suggestion
    protobuf \
```

jnsgruk commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 363626`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>multipass</li>
  </ul>
</details>

getchoo reviewed (approved):
Diff LGTM, but I'm not all that familiar with flutter

ambroisie commented on pkgs/by-name/mu/multipass/gui.nix:
```diff
@@ -0,0 +1,79 @@
+
+    # Generate the Dart gRPC code for the Multipass GUI.
+    protoc \
+      --plugin=protoc-gen-dart=${lib.getExe protoc-gen-dart} \
```
I think as it is, this is broken for cross-compilation? And needs to use either:

* `$(which protoc-gen-dart)` in the shell to look up the `nativeBuildInputs`.
* `${lib.getExe buildPackages.protoc-gen-dart}` in Nix to correctly use a build package, not a host package.
